examples/plot_california_housing.py

```diff
@@ -0,0 +1,1933 @@
+
+# %%
+# As we have already established earlier, we find that the target data
+# seems to be clipped. Let’s take a closer look:
+
+# %%
+df[target_col].value_counts().head()
+
+# %%
+# Is it possible that this would occur naturally in the data? Sometimes,
+# we may find strange patterns. To give an example, if there was a law
+# that sales above a certain value are taxed differently, we could expect
+# prices to cluster at this value. But this appears to be very unlikely
+# here, especially since prices seem to be rounded to the closest $100, as
+# we can see from the following probe:
+
+# %%
+(
+    (df[target_col] % 100)
+    .round()
+    .value_counts()
+    .to_frame()
+    .reset_index()
+    .rename(columns={"index": f"{target_col} % $100", target_col: "count"})
+)
+
+# %%
+# So if we take the modulo of the house price to $100, we find that it it’s
+# almost always 0 (well, we see a few 100’s, but that’s just a rounding issue).
+# Then we see 965 prices ending in , which are exactly those 965 samples we
+# found with a price of $100,001. Finally, we see 4 prices ending in 9. Let’s
+# take a look:
+
+# %%
+df[np.isclose(df[target_col] % 100, 99)][target_col].to_frame()
+
+# %%
+df[target_col].min()
+
+# %%
+# So these four samples actually correspond to the lowest prices in the dataset.
+# Therefore, it is very reasonable to assume that the dataset creators decided
+# to set a maximum price of $500,000 and a minimum price of $15,000, with all
+# prices falling outside that range being set to the max/min price +/- . For the
+# prices within the range, they decided to round to $100.
+
+# %%
+# When it comes to our task of predicting the target, this clipping could
+# be dangerous. We cannot know by how much the actual price was clipped,
+# especially when it comes to high prices. For a machine learning model,
+# it could become extraordinarily hard to predict these high prices,
+# because even though the features may, for instance, indicate a price of
+# $1,000,000 for one sample and $500,100 for another, the model is supposed to
+# predict the same price for both. For this reason, let us remove the
+# clipped data from the dataset. Whether that’s a good idea will in
+# reality depend on the use case that we’re trying to solve.
+
+# %%
+# Training a machine learning model
+# ---------------------------------
+
+# %%
+# Now that we have gained a good understanding of our data, we move to the
+# next phase, which is training a machine learning model to predict the
+# target.
+
+# %%
+# Remove samples with clipped target data
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# %%
+# As a first step, we will remove the samples where the target data has
+# been clipped, as discussed above.
+
+
+# %%
+mask = (15_000 <= df[target_col]) & (df[target_col] <= 500_000)
+print(
+    f"Discarding {(1 - mask).sum()} ({100 * (1 - mask.mean()):.1f}%) of rows because"
+    " the target is clipped."
+)
+
+# %%
+# Train/test split
+# ~~~~~~~~~~~~~~~~
```

Comment:
Yes, it would be a bit "more correct" to do it like that. The issue is that the steps taken above all affect the whole dataset (e.g. scaling the target, removing outliers), so if we split first, we would have to always apply the same operation twice.

I'd argue no test information "leaks" into the ML model, but that's a proposition that can't be proven. If you insist, I can change it, but working in this script format is not very convenient, so I'd like to avoid bigger changes if not really necessary :)

Reply:
When you look at the range of the data, find outliers, do scaling and filtering, that's a lot of information leaked already. I have seen this causing performance of the model being artificially high. That's why I think it's important to have them done by only looking at a training set, or having the ones possible, as a part of the `Pipeline`.

Reply:
Yes, I totally agree with that. Especially, if we'd be identifying outliers based on data, it would be incorrect to look at the test data for that. Here I'd argue it's different though: 

1. We're not removing outliers but rather samples whose target was artificially capped in the dataset. This is something that should actually be documented in the dataset description.
2. When it comes to scaling the units to be `$1`, this is totally arbitrary and not necessary for the analysis (initially, I didn't even do it). The reason why I made the change is purely so that results are comparable with others found online, which use this scaling.

I'd say that none of this leaks information and it is not "learned" from the data.

Reply:
Very fair!